pkg/mutate/manage_pod_mutation.go

```diff
@@ -0,0 +1,132 @@
+package mutate
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+
+	offv1alpha1 "github.com/liqotech/liqo/apis/offloading/v1alpha1"
+	liqoconst "github.com/liqotech/liqo/pkg/consts"
+)
+
+// getVirtualNodeToleration returns a new Toleration for the Liqo's virtual-node.
+func getVirtualNodeToleration() corev1.Toleration {
+	return corev1.Toleration{
+		Key:      liqoconst.VirtualNodeTolerationKey,
+		Operator: corev1.TolerationOpExists,
+		Effect:   corev1.TaintEffectNoExecute,
+	}
+}
+
+// createTolerationFromNamespaceOffloading creates a new virtualNodeToleration in case of LocalAndRemotePodOffloadingStrategyType
+// or RemotePodOffloadingStrategyType. In case of LocalPodOffloadingStrategyType returns an empty Toleration.
+func createTolerationFromNamespaceOffloading(strategy offv1alpha1.PodOffloadingStrategyType) corev1.Toleration {
+	var toleration corev1.Toleration
+	switch {
+	case strategy == offv1alpha1.LocalAndRemotePodOffloadingStrategyType, strategy == offv1alpha1.RemotePodOffloadingStrategyType:
+		// The virtual-node toleration must be added.
+		toleration = getVirtualNodeToleration()
+	default:
+		// LocalPodOffloadingStrategyType, default is less clear but safer
+		// No toleration has to be added.
+		toleration = corev1.Toleration{}
+	}
+	return toleration
+}
+
+// createNodeSelectorFromNamespaceOffloading creates the right NodeSelector according to the PodOffloadingStrategy chosen.
+func createNodeSelectorFromNamespaceOffloading(nsoff *offv1alpha1.NamespaceOffloading) corev1.NodeSelector {
+	var nodeSelector corev1.NodeSelector
+	switch {
+	case nsoff.Spec.PodOffloadingStrategy == offv1alpha1.RemotePodOffloadingStrategyType:
+		// To ensure that the pod is not scheduled on local nodes is necessary to add a new NodeSelectorRequirement which
+		// requires the label "liqo.io/type=virtual-node", to every NodeSelectorTerm.
+		nodeSelector = nsoff.Spec.ClusterSelector
+		for i := range nodeSelector.NodeSelectorTerms {
+			nodeSelector.NodeSelectorTerms[i].MatchExpressions = append(nodeSelector.NodeSelectorTerms[i].MatchExpressions,
```

Comment:
Can you add comments on the reason why you are adding this expression to each term?

Reply:
I have modified a little bit the previous comment, I hope it is clearer now.